File: tests/LinuxAndC6678/Core_SysBios/sobel-c6678AndLinux/Code/src/communication.c
/*******************************************************************************
 * Copyright or © or Copr. %%LOWERDATE%% - %%UPPERDATE%% IETR/INSA:
 *
 * %%AUTHORS%%
 *
 * This software is a computer program whose purpose is to prototype
 * parallel applications.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 *******************************************************************************/
/*
 * communication.c
 *
 *  Created on: 7 août 2013
 *      Author: Karol
 */
#include <xdc/std.h>
/*  ----------------------------------- IPC module Headers           */
#include <ti/ipc/Ipc.h>
#include <ti/ipc/Notify.h>
#include <ti/sysbios/knl/Semaphore.h>
#include <xdc/runtime/System.h>
#include "communication.h"
#include <ti/sysbios/BIOS.h>
#include <ti/csl/csl_cacheAux.h>
#include <ti/csl/csl_semAux.h>
#include <ti/ipc/MultiProc.h>
#include <ti/sysbios/knl/Task.h>
#include <xdc/runtime/Error.h>
#include <ti/ipc/HeapMemMP.h>
#include "com.h"
#define HEAP_NAME  		"HeapMemMP"
// 8 local semaphore for each core (1 useless)
Semaphore_Handle interCoreSem[8];

HeapMemMP_Handle heapHandle;

/* Interrupt line used (0 is default) */
#define INTERRUPT_LINE  0

/* Semaphore used by the busy barrier */
#define BUSY_BARRIER_SEMAPHORE 31


/* Notify event number that the app uses */
#define EVENTID         10

#pragma DATA_SECTION(barrier, ".MSMCSRAM")
Char barrier = 0x01;

Void callbackInterCoreCom(UInt16 procId, UInt16 lineId, UInt32 eventId,
		UArg arg, UInt32 payload) {
	Semaphore_post(interCoreSem[procId]);
}

void sendStart(Uint16 coreID) {
	/* Send an event to the next processor */
	// The last parameter (TRUE) may cause this function to block.
	// In such case, maybe the penultimate parameter could be used to signal
	// that several send are grouped within a single notify event
	Int status = Notify_sendEvent(coreID, INTERRUPT_LINE, EVENTID, 0, TRUE);
	if (status < 0) {
		System_abort("sendEvent failed\n");
	}
}

void receiveEnd(Uint16 coreID) {
	Semaphore_pend(interCoreSem[coreID], BIOS_WAIT_FOREVER);
}

void communicationInit() {
	int i;
	Char procNumber = MultiProc_self();

	/*
	 *  Ipc_start() calls Ipc_attach() to synchronize all remote processors
	 *  because 'Ipc.procSync' is set to 'Ipc.ProcSync_ALL' in *.cfg
	 */
	int status = Ipc_start();
	if (status < 0) {
		System_abort("Ipc_start failed\n");
	}

	Error_init(NULL);

	// Connect to processor 0
	status = 0;
	do {
		status = Ipc_attach(0);
	} while (status < 0);

	heapHandle = openHeap(HEAP_NAME);

	// Allocate and initialize a new semaphores and return its handle
	for (i = 1; i < 8; i++) {
		interCoreSem[i] = Semaphore_create(0, NULL, NULL);
	}

	for (i = 1; i < 8; i++) {
		if (i != procNumber) {
			status = 0;
			do {
				status = Ipc_attach(i);
			} while (status < 0);
		}
	}

	/*
	 *  Register the same callback for all remote processor
	 */
	for (i = 1; i < 8; i++) {
		status = Notify_registerEvent(i, INTERRUPT_LINE, EVENTID,
				(Notify_FnNotifyCbck) callbackInterCoreCom, NULL);
		if (status < 0) {
			System_abort("Notify_registerEvent failed\n");
		}
	}

	if(procNumber == 1){
		hSEM->SEM_RST_RUN = 0x01;
	}
}

void receiveStart() {
}
void sendEnd() {
}

void busy_barrier() {
	Uint8 status;
	Char procNumber = MultiProc_self();

	do {
		status = CSL_semAcquireDirect(BUSY_BARRIER_SEMAPHORE);
	} while (status == 0);

	CACHE_invL2(&barrier, 1, CACHE_WAIT);
	barrier |= (1 << procNumber);
	CACHE_wbInvL2(&barrier, 1, CACHE_WAIT);
	CSL_semReleaseSemaphore(BUSY_BARRIER_SEMAPHORE);

	if (procNumber == 1) {
		while (barrier != (Char) 0xFF) {
			Task_sleep(1);
			CACHE_invL2(&barrier, 1, CACHE_WAIT);
		}
		barrier = (Char) 0x01;
		CACHE_wbInvL2(&barrier, 1, CACHE_WAIT);
		sendStart(2);
		receiveEnd(7);
	} else {
		int sendIdx;
		receiveEnd(procNumber - 1);
		sendIdx = (procNumber + 1) % 8;
		sendIdx = (sendIdx == 0) ? 1 : sendIdx;
		sendStart(sendIdx);

	}
}

File: plugins/org.ietr.preesm.codegen.xtend/ecore-gen/org/ietr/preesm/codegen/model/codegen/Communication.java
/*******************************************************************************
 * Copyright or © or Copr. 2013 - 2017 IETR/INSA:
 *
 * Antoine Morvan <antoine.morvan@insa-rennes.fr> (2017)
 * Karol Desnos <karol.desnos@insa-rennes.fr> (2013)
 *
 * This software is a computer program whose purpose is to prototype
 * parallel applications.
 *
 * This software is governed by the CeCILL-C license under French law and
 * abiding by the rules of distribution of free software.  You can  use
 * modify and/ or redistribute the software under the terms of the CeCILL-C
 * license as circulated by CEA, CNRS and INRIA at the following URL
 * "http://www.cecill.info".
 *
 * As a counterpart to the access to the source code and  rights to copy,
 * modify and redistribute granted by the license, users are provided only
 * with a limited warranty  and the software's author,  the holder of the
 * economic rights,  and the successive licensors  have only  limited
 * liability.
 *
 * In this respect, the user's attention is drawn to the risks associated
 * with loading,  using,  modifying and/or developing or reproducing the
 * software by the user in light of its specific status of free software,
 * that may mean  that it is complicated to manipulate,  and  that  also
 * therefore means  that it is reserved for developers  and  experienced
 * professionals having in-depth computer knowledge. Users are therefore
 * encouraged to load and test the software's suitability as regards their
 * requirements in conditions enabling the security of their systems and/or
 * data to be ensured and,  more generally, to use and operate it in the
 * same conditions as regards security.
 *
 * The fact that you are presently reading this means that you have had
 * knowledge of the CeCILL-C license and that you accept its terms.
 *******************************************************************************/
package org.ietr.preesm.codegen.model.codegen;

import org.eclipse.emf.common.util.EList;

// TODO: Auto-generated Javadoc
/**
 * <!-- begin-user-doc --> A {@link Communication} is a {@link Call} used to represent the transmission of data between
 * processing elements. Each {@link Communication} {@link Call} has a {@link Direction} and a {@link Delimiter}.<!--
 * end-user-doc -->
 *
 * <p>
 * The following features are supported:
 * </p>
 * <ul>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getDirection <em>Direction</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getDelimiter <em>Delimiter</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getData <em>Data</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendStart <em>Send Start</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendEnd <em>Send End</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveStart <em>Receive Start</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveEnd <em>Receive End</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getId <em>Id</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getNodes <em>Nodes</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveRelease <em>Receive Release</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendReserve <em>Send Reserve</em>}</li>
 * <li>{@link org.ietr.preesm.codegen.model.codegen.Communication#isRedundant <em>Redundant</em>}</li>
 * </ul>
 *
 * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication()
 * @model
 * @generated
 */
public interface Communication extends Call {
  /**
   * Returns the value of the '<em><b>Direction</b></em>' attribute. The literals are from the enumeration
   * {@link org.ietr.preesm.codegen.model.codegen.Direction}. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Direction</em>' attribute isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Direction</em>' attribute.
   * @see org.ietr.preesm.codegen.model.codegen.Direction
   * @see #setDirection(Direction)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Direction()
   * @model required="true"
   * @generated
   */
  Direction getDirection();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getDirection
   * <em>Direction</em>}' attribute. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Direction</em>' attribute.
   * @see org.ietr.preesm.codegen.model.codegen.Direction
   * @see #getDirection()
   * @generated
   */
  void setDirection(Direction value);

  /**
   * Returns the value of the '<em><b>Delimiter</b></em>' attribute. The literals are from the enumeration
   * {@link org.ietr.preesm.codegen.model.codegen.Delimiter}. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Delimiter</em>' attribute isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Delimiter</em>' attribute.
   * @see org.ietr.preesm.codegen.model.codegen.Delimiter
   * @see #setDelimiter(Delimiter)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Delimiter()
   * @model required="true"
   * @generated
   */
  Delimiter getDelimiter();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getDelimiter
   * <em>Delimiter</em>}' attribute. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Delimiter</em>' attribute.
   * @see org.ietr.preesm.codegen.model.codegen.Delimiter
   * @see #getDelimiter()
   * @generated
   */
  void setDelimiter(Delimiter value);

  /**
   * Returns the value of the '<em><b>Data</b></em>' reference. <!-- begin-user-doc --> {@link Buffer} that is sent or
   * received by the current {@link Communication}. Calling this method is equivalent to calling
   * <code>thisCommunication.getParameters().iterator().next()</code> since the only "Parameter" of a communication is
   * the buffer it transfers. <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Data</em>' reference.
   * @see #setData(Buffer)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Data()
   * @model required="true"
   * @generated
   */
  Buffer getData();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getData <em>Data</em>}'
   * reference. <!-- begin-user-doc -->Set the {@link #getData()} associated to the current {@link Communication}.
   * Calling this method will also modify the {@link Communication#getParameters() parameters}.<!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Data</em>' reference.
   * @see #getData()
   * @generated
   */
  void setData(Buffer value);

  /**
   * Returns the value of the '<em><b>Send Start</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Send Start</em>' reference isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Send Start</em>' reference.
   * @see #setSendStart(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_SendStart()
   * @model transient="true"
   * @generated
   */
  Communication getSendStart();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendStart <em>Send
   * Start</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Send Start</em>' reference.
   * @see #getSendStart()
   * @generated
   */
  void setSendStart(Communication value);

  /**
   * Returns the value of the '<em><b>Send End</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Send End</em>' reference isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Send End</em>' reference.
   * @see #setSendEnd(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_SendEnd()
   * @model transient="true"
   * @generated
   */
  Communication getSendEnd();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendEnd <em>Send
   * End</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Send End</em>' reference.
   * @see #getSendEnd()
   * @generated
   */
  void setSendEnd(Communication value);

  /**
   * Returns the value of the '<em><b>Receive Start</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Receive Start</em>' reference isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Receive Start</em>' reference.
   * @see #setReceiveStart(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_ReceiveStart()
   * @model transient="true"
   * @generated
   */
  Communication getReceiveStart();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveStart <em>Receive
   * Start</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Receive Start</em>' reference.
   * @see #getReceiveStart()
   * @generated
   */
  void setReceiveStart(Communication value);

  /**
   * Returns the value of the '<em><b>Receive End</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Receive End</em>' reference isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Receive End</em>' reference.
   * @see #setReceiveEnd(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_ReceiveEnd()
   * @model transient="true"
   * @generated
   */
  Communication getReceiveEnd();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveEnd <em>Receive
   * End</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Receive End</em>' reference.
   * @see #getReceiveEnd()
   * @generated
   */
  void setReceiveEnd(Communication value);

  /**
   * Returns the value of the '<em><b>Id</b></em>' attribute. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Id</em>' attribute isn't clear, there really should be more of a description here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Id</em>' attribute.
   * @see #setId(int)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Id()
   * @model required="true"
   * @generated
   */
  int getId();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getId <em>Id</em>}'
   * attribute. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Id</em>' attribute.
   * @see #getId()
   * @generated
   */
  void setId(int value);

  /**
   * Returns the value of the '<em><b>Nodes</b></em>' containment reference list. The list contents are of type
   * {@link org.ietr.preesm.codegen.model.codegen.CommunicationNode}. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Nodes</em>' containment reference list isn't clear, there really should be more of a
   * description here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Nodes</em>' containment reference list.
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Nodes()
   * @model containment="true" required="true"
   * @generated
   */
  EList<CommunicationNode> getNodes();

  /**
   * Returns the value of the '<em><b>Receive Release</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Receive Release</em>' reference isn't clear, there really should be more of a
   * description here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Receive Release</em>' reference.
   * @see #setReceiveRelease(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_ReceiveRelease()
   * @model transient="true"
   * @generated
   */
  Communication getReceiveRelease();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getReceiveRelease
   * <em>Receive Release</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Receive Release</em>' reference.
   * @see #getReceiveRelease()
   * @generated
   */
  void setReceiveRelease(Communication value);

  /**
   * Returns the value of the '<em><b>Send Reserve</b></em>' reference. <!-- begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Send Reserve</em>' reference isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   * 
   * @return the value of the '<em>Send Reserve</em>' reference.
   * @see #setSendReserve(Communication)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_SendReserve()
   * @model transient="true"
   * @generated
   */
  Communication getSendReserve();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#getSendReserve <em>Send
   * Reserve</em>}' reference. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Send Reserve</em>' reference.
   * @see #getSendReserve()
   * @generated
   */
  void setSendReserve(Communication value);

  /**
   * Returns the value of the '<em><b>Redundant</b></em>' attribute. The default value is <code>"false"</code>. <!--
   * begin-user-doc -->
   * <p>
   * If the meaning of the '<em>Redundant</em>' attribute isn't clear, there really should be more of a description
   * here...
   * </p>
   * <!-- end-user-doc -->
   *
   * @return the value of the '<em>Redundant</em>' attribute.
   * @see #setRedundant(boolean)
   * @see org.ietr.preesm.codegen.model.codegen.CodegenPackage#getCommunication_Redundant()
   * @model default="false"
   * @generated
   */
  boolean isRedundant();

  /**
   * Sets the value of the '{@link org.ietr.preesm.codegen.model.codegen.Communication#isRedundant
   * <em>Redundant</em>}' attribute. <!-- begin-user-doc --> <!-- end-user-doc -->
   * 
   * @param value
   *          the new value of the '<em>Redundant</em>' attribute.
   * @see #isRedundant()
   * @generated
   */
  void setRedundant(boolean value);

  /**
   * <!-- begin-user-doc --> This method browse the {@link #eContainer()} recursively of the {@link Communication} until
   * it finds a {@link CoreBlock}. <!-- end-user-doc -->
   * 
   * @model kind="operation" required="true"
   * @generated
   */
  CoreBlock getCoreContainer();

} // Communication
